core/state/trie_prefetcher.go

```diff
@@ -248,40 +205,47 @@ func newSubfetcher(db Database, state common.Hash, owner common.Hash, root commo
 }
 
 // schedule adds a batch of trie keys to the queue to prefetch.
-func (sf *subfetcher) schedule(keys [][]byte) {
+func (sf *subfetcher) schedule(keys [][]byte) error {
 	// Append the tasks to the current queue
 	sf.lock.Lock()
 	sf.tasks = append(sf.tasks, keys...)
 	sf.lock.Unlock()
 
-	// Notify the prefetcher, it's fine if it's already terminated
+	// Notify the background thread to execute scheduled tasks
 	select {
-	case sf.wake <- struct{}{}:
+	case <-sf.term:
+		// Reject scheduling a task if the fetcher's already terminated
+		return errTerminated
 	default:
+		// Otherwise notify the fetcher if it's not already done so (async!)
+		//
+		// Note, the default branch is to give priority to term check as the
+		// select evaluates branches randomly.
+		select {
+		case sf.wake <- struct{}{}:
+			// Wake signal sent
+		default:
+			// Wake signal not sent as a previous is already queued
+		}
+		return nil
 	}
 }
 
-// peek tries to retrieve a deep copy of the fetcher's trie in whatever form it
-// is currently.
+// peek tries to retrieve a deep copy of the fetcher's trie. Nil is returned
+// if the fetcher is already terminated, or the associated trie is failing
+// for opening.
 func (sf *subfetcher) peek() Trie {
 	ch := make(chan Trie)
 	select {
 	case sf.copy <- ch:
-		// Subfetcher still alive, return copy from it
 		return <-ch
-
 	case <-sf.term:
-		// Subfetcher already terminated, return a copy directly
-		if sf.trie == nil {
-			return nil
-		}
-		return sf.db.CopyTrie(sf.trie)
+		return nil
 	}
 }
 
-// abort interrupts the subfetcher immediately. It is safe to call abort multiple
-// times but it is not thread safe.
-func (sf *subfetcher) abort() {
+// close waits for the subfetcher to finish its tasks. It cannot be called multiple times
```

Comment:
But it can be called multiple times. Close might also not be the best name since we're waiting for it to finish but should AFAIK not kill the thing.